File: src/sorting/randomized_select.cpp
// randomized_select.cpp — CLRS 4e §9.2：期望線性時間找第 i 小
#include <iostream>
#include <random>
#include <vector>

int partition(std::vector<int>& A, int p, int r) {
    int x = A[r], i = p - 1;
    for (int j = p; j < r; ++j)
        if (A[j] <= x) std::swap(A[++i], A[j]);
    std::swap(A[i + 1], A[r]);
    return i + 1;
}

// 回傳 A[p..r] 中第 i 小（i 從 1 起算）
int randomizedSelect(std::vector<int>& A, int p, int r, int i, std::mt19937& rng) {
    if (p == r) return A[p];
    std::uniform_int_distribution<int> d(p, r);
    std::swap(A[d(rng)], A[r]);
    int q = partition(A, p, r);
    int k = q - p + 1;                  // pivot 在這段裡排第 k 小
    if (i == k) return A[q];            // 運氣好：pivot 就是答案
    if (i < k) return randomizedSelect(A, p, q - 1, i, rng);          // 只走一邊！
    return randomizedSelect(A, q + 1, r, i - k, rng);
}

int main() {
    std::vector<int> A = {3, 2, 9, 0, 7, 5, 4, 8, 6, 1};
    std::mt19937 rng(9);
    for (int i : {1, 5, 10}) {
        std::vector<int> tmp = A;
        std::cout << "第 " << i << " 小 = "
                  << randomizedSelect(tmp, 0, static_cast<int>(tmp.size()) - 1, i, rng) << '\n';
    }
    return 0;
}

File: src/sorting/randomized_select.py
# 改寫自官方 Chapter 9/randomized_select.py（MIT License）：含 partition 全文
import random

random.seed(42)


def partition(A, p, r):
    x = A[r]
    i = p - 1
    for j in range(p, r):
        if A[j] <= x:
            i += 1
            A[i], A[j] = A[j], A[i]
    A[i + 1], A[r] = A[r], A[i + 1]
    return i + 1


def randomized_partition(A, p, r):
    i = random.randint(p, r)
    A[i], A[r] = A[r], A[i]
    return partition(A, p, r)


def randomized_select(A, p, r, i):
    if p == r:
        return A[p]
    q = randomized_partition(A, p, r)
    k = q - p + 1                       # pivot 在本段排第 k 小
    if i == k:
        return A[q]
    if i < k:
        return randomized_select(A, p, q - 1, i)
    return randomized_select(A, q + 1, r, i - k)   # 進右半段要扣掉左邊 k 個


A = [3, 2, 9, 0, 7, 5, 4, 8, 6, 1]
for i in (1, 5, 10):
    print("第", i, "小 =", randomized_select(A[:], 0, len(A) - 1, i))
